File: data_structures/heap/max_heap.dart
// max_heap.dart

class MaxHeap {
  List<int> _heap;

  MaxHeap() : _heap = [];

  void insert(int value) {
    _heap.add(value);
    _heapifyUp();
  }

  void _heapifyUp() {
    int index = _heap.length - 1;

    while (index > 0) {
      int parentIndex = (index - 1) ~/ 2;

      if (_heap[index] > _heap[parentIndex]) {
        _swap(index, parentIndex);
        index = parentIndex;
      } else {
        break;
      }
    }
  }

  int extractMax() {
    if (_heap.isEmpty) {
      throw StateError("Heap is empty");
    }

    int maxValue = _heap[0];

    if (_heap.length == 1) {
      _heap.removeAt(0);
    } else {
      _heap[0] = _heap.removeLast();
      _heapifyDown();
    }

    return maxValue;
  }

  void _heapifyDown() {
    int index = 0;

    while (true) {
      int leftChildIndex = 2 * index + 1;
      int rightChildIndex = 2 * index + 2;
      int largestChildIndex = index;

      if (leftChildIndex < _heap.length &&
          _heap[leftChildIndex] > _heap[largestChildIndex]) {
        largestChildIndex = leftChildIndex;
      }

      if (rightChildIndex < _heap.length &&
          _heap[rightChildIndex] > _heap[largestChildIndex]) {
        largestChildIndex = rightChildIndex;
      }

      if (largestChildIndex != index) {
        _swap(index, largestChildIndex);
        index = largestChildIndex;
      } else {
        break;
      }
    }
  }

  void _swap(int i, int j) {
    int temp = _heap[i];
    _heap[i] = _heap[j];
    _heap[j] = temp;
  }

  void display() {
    print("Max Heap: $_heap");
  }
}

void main() {
  MaxHeap maxHeap = MaxHeap();

  maxHeap.insert(3);
  maxHeap.insert(1);
  maxHeap.insert(4);
  maxHeap.insert(2);

  print("Original Max Heap:");
  maxHeap.display();

  print("\nExtract Max Value: ${maxHeap.extractMax()}");

  print("Max Heap after Extraction:");
  maxHeap.display();
}

File: go/data_structures/heap/max_heap.go
// max_heap.go

package main

import "fmt"

type MaxHeap struct {
    heap []int
}

func NewMaxHeap() *MaxHeap {
    return &MaxHeap{}
}

func (h *MaxHeap) Insert(value int) {
    h.heap = append(h.heap, value)
    h.heapifyUp(len(h.heap) - 1)
}

func (h *MaxHeap) ExtractMax() (int, error) {
    if len(h.heap) == 0 {
        return 0, fmt.Errorf("heap is empty")
    }

    max := h.heap[0]
    lastIdx := len(h.heap) - 1

    h.heap[0], h.heap[lastIdx] = h.heap[lastIdx], h.heap[0]
    h.heap = h.heap[:lastIdx]

    h.heapifyDown(0)

    return max, nil
}

func (h *MaxHeap) heapifyUp(index int) {
    parent := (index - 1) / 2

    for index > 0 && h.heap[index] > h.heap[parent] {
        h.heap[index], h.heap[parent] = h.heap[parent], h.heap[index]
        index = parent
        parent = (index - 1) / 2
    }
}

func (h *MaxHeap) heapifyDown(index int) {
    for {
        leftChild := 2*index + 1
        rightChild := 2*index + 2
        largest := index

        if leftChild < len(h.heap) && h.heap[leftChild] > h.heap[largest] {
            largest = leftChild
        }

        if rightChild < len(h.heap) && h.heap[rightChild] > h.heap[largest] {
            largest = rightChild
        }

        if largest != index {
            h.heap[index], h.heap[largest] = h.heap[largest], h.heap[index]
            index = largest
        } else {
            break
        }
    }
}

func main() {
    heap := NewMaxHeap()

    heap.Insert(4)
    heap.Insert(10)
    heap.Insert(3)
    heap.Insert(5)
    heap.Insert(1)

    fmt.Println("Max Heap:", heap.heap)

    max, err := heap.ExtractMax()
    if err == nil {
        fmt.Printf("Extracted Max Element: %d\n", max)
        fmt.Println("Updated Max Heap:", heap.heap)
    } else {
        fmt.Println(err)
    }
}
